### ISSUE TITLE
How mean pooling actually works?

### ISSUE THREAD

**Issue by Charlie-Lin** [UKPLab/sentence-transformers#230]
Hello!
Thank you for publishing this great package for everyone !

In "Pooling.py", I can't understand the concept behind doing 
**"sum_embeddings = torch.sum(token_embeddings * input_mask_expanded, 1)"**.  
https://github.com/UKPLab/sentence-transformers/blob/master/sentence_transformers/models/Pooling.py

![image](https://user-images.githubusercontent.com/22322921/82014027-48518780-96ae-11ea-842f-2567df1bf96e.png)

Does it mean that the padding part will be masked while doing mean pooling?

e.g. 
`token_embeddings  = [1, 2, 3, 0, 0, 0, 0]   #The last four are padding tokens`
`input_mask_expanded = [1, 1, 1, 0, 0, 0, 0]`
`sum_embeddings  = [6]   #6 = 1*1 + 2*1 + 3*1`

**Comment by nreimers** [comment]
Correct, padding tokens will be ignored for the computation of the mean. Otherwise the mean of a sentence would influenced by the padding length, i.e. Hello World PAD PAD would yield a different sentence embedding than Hello World

**Comment by Charlie-Lin** [comment]
Thanks for your prompt reply. 
I can realize the difference between masked and non-masked padding tokens in calculation. 

However, in my understanding, BERT Model project input tokens onto high dimensional space through bert layers (e.g. D=768), and output an embedding vector at the end like :
`input_tokens = [1, 2, 3, 0, 0, 0] `
`output_embedding = [0.2, -1, 0.3, 0.5, -0.1, 0,7]` 

Therefore, after doing mean pooling, the result of **masking** padding token is:
`[1*0,2 + 1*(-1) + 1*0,3 + 0*0,5 + 0*(-0,1) +0*0,7] / 3 = -0.3`

and the result of **non-masking** padding token is:
`[0,2 + (-1) + 0,3 + 0,5 + (-0,1) +0,7] / 6 = 0.1`

Will the result of  **masking** padding token lose more latent information?
